server/authEndpoint.js

```diff
@@ -0,0 +1,192 @@
+const { Pool } = require("pg");
+const jwt = require("jsonwebtoken");
+const { check, validationResult } = require("express-validator");
+const cors = require("cors");
+const bcrypt = require("bcrypt");
+const secret = "mysecret";
+
+const pool = new Pool({
+	user: "postgres",
+	host: "localhost",
+	database: "study_buddies",
+	password: "CYFStudent123",
+	port: 5432,
+});
+
+app.use(
+	cors({
+		accessControlAllowOrigin: "*",
+	})
+);
+
+// Provide users to create an account by providing their email, password, and name.
+app.post(
+	"/register",
+	[
+		check("email", "Please include a valid email").isEmail(),
+		check(
+			"password",
+			"Please enter a password with 6 or more characters"
+		).isLength({ min: 6 }),
+		check("name", "Please enter a name").isLength({ min: 1 }),
+	],
+	async (req, res) => {
+		const errors = validationResult(req);
+		if (!errors.isEmpty()) {
+			return res.status(400).json({ errors: errors.array() });
+		}
+
+		const { email, password, name } = req.body;
+
+		try {
+			const salt = await bcrypt.genSalt(10);
+			const hashedPassword = await bcrypt.hash(password, salt);
+			const result = await pool.query(
+				"SELECT * FROM trainees WHERE email = $1",
+				[email]
+			);
+			if (result.rows.length > 0) {
+				return res
+					.status(400)
+					.json({ errors: [{ msg: "User already exists" }] });
+			}
+			const query =
+				"INSERT INTO trainees (email, password, name) VALUES ($1, $2, $3)";
+			const values = [email, hashedPassword, name];
+			await pool.query(query, values);
+			res.status(201).json({ msg: "User created" });
+		} catch (err) {
+			//eslint-disable-next-line
+			console.error(err);
+			res.status(500).json({ errors: [{ msg: "Server error" }] });
+		}
+	}
+);
+
+// Provide allows registered users to log in by providing their email and password.
+// If the credentials are valid, the server returns a JWT that can be used to authenticate subsequent requests.
+app.post("/login", async (req, res) => {
+	const errors = validationResult(req);
+	if (!errors.isEmpty()) {
+		return res.status(400).json({ errors: errors.array() });
+	}
+
+	const { email, password } = req.body;
+
+	try {
+		const result = await pool.query("SELECT * FROM trainees WHERE email = $1", [
+			email,
+		]);
+		if (result.rows.length === 0) {
+			return res.status(400).json({ errors: [{ msg: "Invalid credentials" }] });
+		}
+
+		const user = result.rows[0];
+		const isMatch = await bcrypt.compare(password, user.password);
+		if (!isMatch) {
+			return res.status(400).json({ errors: [{ msg: "Invalid credentials" }] });
+		}
+
+		const token = jwt.sign({ userId: user.id }, secret, { expiresIn: "1h" });
+		res.json({ token });
+	} catch (err) {
+		//eslint-disable-next-line
+		console.error(err);
+		res.status(500).json({ errors: [{ msg: "Server error" }] });
+	}
+});
+
+// Returns a list of all availability records in the database.
+// Only authenticated users can access this endpoint.
+app.get("/availabilities/", (req, res) => {
```

Comment:
const queryOptions = {
        daily: `WHERE availability_date = CURRENT_DATE`,
        weekly: `WHERE availability_date >= CURRENT_DATE - interval '7' day`,
        monthly: `WHERE date_trunc('month', availability_date) = date_trunc('month', CURRENT_DATE)`,
        name: `WHERE t.name = '${req.query.name}'`,
        email: `WHERE t.email = '${req.query.email}'`,
    };
    const queryFilter = queryOptions[req.query.filter] || '';
    const search = req.query.search || '';
    const query = `
        SELECT a.availability_date, a.topic, t.name, t.email
        FROM availability a
                 JOIN trainees t ON a.trainees_id = t.id ${queryFilter}
                AND (a.topic ILIKE '%${search}%' OR t.name ILIKE '%${search}%' OR t.email ILIKE '%${search}%')
        ORDER BY a.availability_date DESC
          
    `;

    const {rows} = await pool.query(query);
    res.json({
        data: rows,
    });

Reply:
👍🏻 Thank you Ender.